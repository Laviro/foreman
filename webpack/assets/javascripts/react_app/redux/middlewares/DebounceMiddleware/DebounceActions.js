import {
  DEBOUNCE_START,
  DEBOUNCE_CLEAR,
  DEBOUNCE_STOP_INCOMING_ACTION,
} from './DebounceConstants';

export const startDebounce = ({ key, debounceID }) => ({
  type: DEBOUNCE_START,
  payload: {
    key,
    debounceID,
  },
});

export const clearDebounce = key => ({
  type: DEBOUNCE_CLEAR,
  payload: {
    key,
  },
});

export const stopIncomingAction = key => ({
  type: DEBOUNCE_STOP_INCOMING_ACTION,
  payload: {
    key,
  },
});
