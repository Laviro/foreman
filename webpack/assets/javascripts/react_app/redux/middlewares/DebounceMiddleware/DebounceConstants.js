export const DEBOUNCE_START = 'DEBOUNCE_START';
export const DEBOUNCE_CLEAR = 'DEBOUNCE_CLEAR';
export const DEBOUNCE_STOP_INCOMING_ACTION = 'DEBOUNCE_STOP_INCOMING_ACTION';
export const DEFAULT_DEBOUNCE = 200;
