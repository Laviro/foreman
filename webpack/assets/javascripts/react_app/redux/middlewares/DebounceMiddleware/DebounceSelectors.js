export const selectDebounce = state => state.debounce || {};

export const selectDebounceItem = (state, key) => selectDebounce(state)[key];
