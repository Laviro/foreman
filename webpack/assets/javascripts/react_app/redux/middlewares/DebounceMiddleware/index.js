export { DebounceMiddleware } from './DebounceMiddleware';
export { reducers } from './DebounceReducer';
export { startDebounce, clearDebounce } from './DebounceActions';
export { withDebounce } from './DebounceHelpers';
