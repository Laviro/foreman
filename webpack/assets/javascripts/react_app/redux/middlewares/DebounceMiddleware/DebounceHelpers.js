import { DEFAULT_DEBOUNCE } from './DebounceConstants';

export const withDebounce = (action, debounce = DEFAULT_DEBOUNCE) => ({
  ...action,
  debounce,
});
