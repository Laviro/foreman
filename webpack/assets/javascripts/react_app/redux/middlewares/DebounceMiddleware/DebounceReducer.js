import Immutable from 'seamless-immutable';
import { DEBOUNCE_START, DEBOUNCE_CLEAR } from './DebounceConstants';

const initialState = Immutable({});

export const reducer = (state = initialState, action) => {
  const { type, payload: { key, debounceID } = {} } = action;
  switch (type) {
    case DEBOUNCE_START:
      return state.merge({ [key]: debounceID });
    case DEBOUNCE_CLEAR:
      return state.without(key);
    default:
      return state;
  }
};

export const reducers = { debounce: reducer };

export default reducer;
