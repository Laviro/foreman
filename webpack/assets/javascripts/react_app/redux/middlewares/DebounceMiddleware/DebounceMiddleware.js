import { omit } from 'lodash';
import { selectDebounceItem } from './DebounceSelectors';
import {
  startDebounce,
  clearDebounce,
  stopIncomingAction,
} from './DebounceActions';

export const DebounceMiddleware = store => next => action => {
  const { key, debounce, payload = {} } = action;
  const debounceKey = key || payload.key;
  const state = store.getState();

  if (!debounce) {
    return next(action);
  }

  if (selectDebounceItem(state, debounceKey)) {
    return store.dispatch(stopIncomingAction(key));
  }

  const modifiedAction = omit(action, ['debounce']);
  return store.dispatch(
    startDebounce({
      key: debounceKey,
      debounceID: setTimeout(() => {
        store.dispatch(modifiedAction);
        store.dispatch(clearDebounce(debounceKey));
      }, debounce),
    })
  );
};

export default DebounceMiddleware;
