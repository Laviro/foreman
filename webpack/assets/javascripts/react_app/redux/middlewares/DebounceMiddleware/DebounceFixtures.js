import { DEFAULT_DEBOUNCE } from './DebounceConstants';
import { withDebounce } from './DebounceHelpers';

export const type = 'SOME_TYPE';
export const key = 'SOME_KEY';
export const payload = { test: true };
export const debounce = DEFAULT_DEBOUNCE;
export const debounceID = 121;
export const initialState = {};
export const stateWithKey = { [key]: debounceID };
export const action = { type, key, payload };
export const actionWithDebounce = withDebounce(action);
