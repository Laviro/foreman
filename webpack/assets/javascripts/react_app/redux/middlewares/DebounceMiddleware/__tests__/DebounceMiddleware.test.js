import { DebounceMiddleware } from '../DebounceMiddleware';
import { clearDebounce } from '../DebounceActions';
import {
  key,
  initialState,
  stateWithKey,
  actionWithDebounce,
} from '../DebounceFixtures';

describe('Debounce Middleware', () => {
  const getFakeStore = () => ({
    getState: () => ({
      debounce: initialState,
    }),
    dispatch: jest.fn(),
  });

  const getFakeStoreWithKey = () => ({
    getState: () => ({
      debounce: stateWithKey,
    }),
    dispatch: jest.fn(),
  });

  it('should handle an action with "debounce" key', () => {
    const fakeStore = getFakeStore();
    const fakeNext = jest.fn();

    DebounceMiddleware(fakeStore)(fakeNext)(actionWithDebounce);
    expect(fakeStore.dispatch).toMatchSnapshot();
    expect(fakeNext).not.toBeCalled();
  });

  it('should stop incoming action due to active debounce', () => {
    const fakeStore = getFakeStoreWithKey();
    const fakeNext = jest.fn();

    DebounceMiddleware(fakeStore)(fakeNext)(actionWithDebounce);
    expect(fakeStore.dispatch).toMatchSnapshot();
  });

  it('should handle DEBOUNCE_CLEAR action', () => {
    const fakeStore = getFakeStoreWithKey();
    const fakeNext = jest.fn();
    const stopAction = clearDebounce(key);

    DebounceMiddleware(fakeStore)(fakeNext)(stopAction);
    expect(fakeNext).toMatchSnapshot();
  });
});
