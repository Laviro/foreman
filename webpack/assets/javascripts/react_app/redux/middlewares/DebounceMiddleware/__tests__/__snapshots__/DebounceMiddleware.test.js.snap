// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Debounce Middleware should handle DEBOUNCE_CLEAR action 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": Object {
          "key": "SOME_KEY",
        },
        "type": "DEBOUNCE_CLEAR",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Debounce Middleware should handle an action with "debounce" key 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": Object {
          "debounceID": 2,
          "key": "SOME_KEY",
        },
        "type": "DEBOUNCE_START",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;

exports[`Debounce Middleware should stop incoming action due to active debounce 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      Object {
        "payload": Object {
          "key": "SOME_KEY",
        },
        "type": "DEBOUNCE_STOP_INCOMING_ACTION",
      },
    ],
  ],
  "results": Array [
    Object {
      "type": "return",
      "value": undefined,
    },
  ],
}
`;
